lib/active_model_serializers/model.rb

```diff
@@ -6,7 +6,36 @@ class Model
     include ActiveModel::Serializers::JSON
     include ActiveModel::Model
 
-    class_attribute :attribute_names
+    # Configuration to avoid a breaking change with older versions of this class which lacked defined attributes.
+    # Previous behavior was: 1) initialized attributes were the
+    class_attribute :attributes_are_always_the_initialization_data, instance_writer: false, instance_reader: false
+    self.attributes_are_always_the_initialization_data = false # remove this commit, just for demonstration
+    module AttributesAreAlwaysTheInitializationData
+      def initialize(attributes = {})
+        @initialized_attributes = attributes && attributes.symbolize_keys
+        super
+      end
+
+      # Defaults to the downcased model name.
+      def id
+        @initialized_attributes.fetch(:id) { self.class.model_name.name && self.class.model_name.name.downcase }
+      end
+
+      def attributes
+        @initialized_attributes
+      end
+    end
+
+    def self.inherited(subclass)
+      if subclass.attributes_are_always_the_initialization_data
+        unless subclass.included_modules.include?(AttributesAreAlwaysTheInitializationData)
+          subclass.prepend(AttributesAreAlwaysTheInitializationData)
```

Comment:
@NullVoxPopuli this is just a compatibility layer to avoid a breaking change.  I should add some docs around setting it true vs. false and how to do both at same time and why :)